cognitedata/reveal
# refactor: recalculateBoundingBox() is only executed when 3D models are added/removed/changed

#### Type of change

![Refactor](https://img.shields.io/badge/Type-Refactor-lightgrey) 

#### Jira ticket :blue_book:


https://cognitedata.atlassian.net/browse/REV-598

## Description :pencil:


 - **Note: The result from `computeBoundsOfMostGeometry` is already cached so this should not effect performance. See code from SectorScene.ts below.**

- Removed `recalculateBoundingBox` from animate.

- Renamed `recalculateBoundingBox` to `forceUpdateCameraNearAndFarBoundingBox` and made it public.

- Added calls to `forceUpdateCameraNearAndFarBoundingBox` everywhere we add or remove models to the viewer.

- Updated 'Reveal exampes' to call `forceUpdateCameraNearAndFarBoundingBox` when there are changes to models.

Cached `computeBoundsOfMostGeometry`:
```
getBoundsOfMostGeometry(): THREE.Box3 {
    this._cachedBoundsOfMostGeometry = this._cachedBoundsOfMostGeometry ?? this.computeBoundsOfMostGeometry();
    return this._cachedBoundsOfMostGeometry;
  }
```

[Finally, note that in our Example project gizmos have a huge bounding box, and since this is also included in the forceUpdateCameraNearAndFarBoundingBox the final bounding box will be too large once a gizmo is added.](https://cognitedata.atlassian.net/browse/REV-598?focusedCommentId=222338)

## How has this been tested? :mag:



Using 'Reveal Examples'  I have removed, added and moved CDF models (cad&pc) and GLTF models, while checking that the bounds calculated in recalculateBoundingBox looks correct and are updated when expected.

## Test instructions :information_source:



See above. Node that you need [this change](https://github.com/cognitedata/reveal/pull/2804) for gizmos to work correctly.

## Checklist :ballot_box_with_check:
